### PR TITLE
[DEV-698] - Implement EIP-5192

### PR DESCRIPTION
**Pull request by grp06**
## Describe your changes
In the PR I add support  for EIP-5192 (https://eips.ethereum.org/EIPS/eip-5192)

When a token is minted we emit the `Lock` event.
When a token is burned, we emit an `Unlock` event (rationale discussed here: https://ethereum-magicians.org/t/final-eip-5192-minimal-soulbound-nfts/9814/15)

With our contract, all tokens will always be locked for as long as they exist. So the `locked()` function will always return `true`

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] I have updated the documentation accordingly. (if applicable)
- [ ] All new and existing tests passed.

### CODE REVIEW

**Comment by rahulrumalla** [comment]
Can you share gas report here? How much more gas does this introduce?

**Comment by TimDaub** [comment]
nice :)

File: src/Badges.sol [hunk 1]
```diff
@@ -272,11 +275,15 @@ contract Badges is
...
 
     emit Transfer(address(0), _to, tokenId);
-
+    emit Locked(tokenId);
     specDataHolder.setBadgeToRaft(tokenId, raftTokenId);
     return tokenId;
   }
 
+  function locked(uint256) public pure returns (bool) {
+    return true;
```
Comment (rahulrumalla) [inline src/Badges.sol:284]: This should return true only if the badge exists, no? For example, a burned badge or a non-existant badge will also return true here. 

Can you add test coverage to these 2 cases, please?

**Comment by TimDaub** [comment]
With EIP-5192 you may need to implement the EIP-721 interface and revert on transfers and approves. I guess it's a bit of a matter of interpretation...

**Comment by grp06** [comment]
> Can you share gas report here? How much more gas does this introduce?

Before the `testBalanceIncreaseAfterTake` function was `398383`
now its `421886` = `23,503` difference

**Comment by rahulrumalla** [comment]
> 

Ah! I see. You mean implement the interface and just revert on all the irrelevant implementations of the interface like `Approve` & `SafeTransferFrom` etc, yeah?

**Comment by rahulrumalla** [comment]
> With EIP-5192 you may need to implement the EIP-721 interface and revert on transfers and approves. I guess it's a bit of a matter of interpretation...

Ah! I see. You mean implement the interface and just revert on all the irrelevant implementations of the interface like `Approve` & `SafeTransferFrom` etc, yeah?

**Comment by rahulrumalla** [comment]
> 398383

**Comment by TimDaub** [comment]
> > With EIP-5192 you may need to implement the EIP-721 interface and revert on transfers and approves. I guess it's a bit of a matter of interpretation...
> 
> Ah! I see. You mean implement the interface and just revert on all the irrelevant implementations of the interface like `Approve` & `SafeTransferFrom` etc, yeah?

yeah. Although it could be that not implementing the functions has the same effect as reverting upon calling them (an error).

**Review by rahulrumalla — APPROVED** [review]

**Review by rahulrumalla — CHANGES_REQUESTED** [review]
Any update on the `Locked()` returning true, @grp06 ? Lmk if u were thinking differently here.

File: src/Badges.sol (hunk 1 above)
Comment (grp06) [reply src/Badges.sol:284]: fixed

**Comment by grp06** [comment]
> Any update on the `Locked()` returning true, @grp06 ? Lmk if u were thinking differently here.

Updated.

**Comment by grp06** [comment]
> With EIP-5192 you may need to implement the EIP-721 interface and revert on transfers and approves. I guess it's a bit of a matter of interpretation...

I'm not following the rationale here. @rahulrumalla do you think I should implement this interface and revert on all of these? How will that help us?

**Comment by TimDaub** [comment]
> > With EIP-5192 you may need to implement the EIP-721 interface and revert on transfers and approves. I guess it's a bit of a matter of interpretation...
> 
> I'm not following the rationale here. @rahulrumalla do you think I should implement this interface and revert on all of these? How will that help us?

EIP-5192 expects an EIP-721 contract that reverts on all transfers and approves. That‘s what is written in the specification. I mean I think not implementing those functions may result in the same effect (which is that if you call them, an error is returned). But I don‘t know if there are second order effects that arise from not implementing them. I personally am in favor of better safe than sorry in this case but it‘s your code base and I‘m also happy to have someone experiment with not implementing them.

But e.g. we should male sure supportsinterfsxe returns true for eip-721 identifiers.

**Comment by rahulrumalla** [comment]
@TimDaub implementing 721 just to revert by default would also mean that there's no full adherence to spec, right? 

I mean if `supportsInterface` returns true for 721 but not actually implement it, it might also have some weird interoperability issues is what I'm thinking. Then what would be point of partially adhering to the spec here in the conundrum.

**Comment by rahulrumalla** [comment]
Hmm I think adding this would be a good move also for interoperability with wider ecosystem actually

**Comment by TimDaub** [comment]
> @TimDaub implementing 721 just to revert by default would also mean that there's no full adherence to spec, right?

I don‘t understand this argument. Why should one not be allowed to inplement eip-721 and revert on all transfers?

File: src/Badges.sol [hunk 2]
```diff
@@ -238,10 +250,17 @@ contract Badges is
...
+    public
+    view
+    virtual
+    override(ERC165Upgradeable, IERC165Upgradeable)
+    returns (bool)
+  {
     return
       _interfaceId == type(IERC721Metadata).interfaceId ||
       _interfaceId == type(IERC4973).interfaceId ||
+      _interfaceId == type(IERC5192).interfaceId ||
```
Comment (rahulrumalla) [inline src/Badges.sol:263]: Now that you implemented 5192, we can expressed here that we support 721.

File: src/Badges.sol [hunk 3]
```diff
@@ -238,10 +250,17 @@ contract Badges is
...
+  function supportsInterface(bytes4 _interfaceId)
+    public
+    view
+    virtual
+    override(ERC165Upgradeable, IERC165Upgradeable)
+    returns (bool)
+  {
     return
       _interfaceId == type(IERC721Metadata).interfaceId ||
       _interfaceId == type(IERC4973).interfaceId ||
```
Comment (rahulrumalla) [inline src/Badges.sol:262]: Can we confirm that the interfaceId now matches what's in EIP spec, since we localized it? Do we have a test for this btw?

Comment (rahulrumalla) [reply src/Badges.sol:262]: We verified this off thread. Resolving this.

**Pull request closed without merging**